src/metkit/api/metkit_c.h

```diff
@@ -0,0 +1,238 @@
+ */
+int metkit_free_marsrequest(const metkit_marsrequest_t* request);
+
+/** Add parameter and values to request
+ * @param request Request instance
+ * @param param parameter name
+ * @param values array of values for parameter
+ * @param numValues number of values
+ * @return int Error code
+ */
+int metkit_marsrequest_add(metkit_marsrequest_t* request, const char* param, const char* values[], int numValues);
+
+/** Add parameter and values to request
+ * @param request Request instance
+ * @param param parameter name
+ * @param values value to add
+ * @return int Error code
+ */
+int metkit_marsrequest_add1(metkit_marsrequest_t* request, const char* param, const char* value);
+
+/** Set verb in Request object
+ * @param request Request instance
+ * @param verb verb to set
+ * @return int Error code
+ */
+int metkit_marsrequest_set_verb(metkit_marsrequest_t* request, const char* verb);
+
+/** Returns the verb in Request object
+ * @param request Request instance
+ * @param verb verb in request
+ * @return int Error code
+ */
+int metkit_marsrequest_verb(const metkit_marsrequest_t* request, const char** verb);
+
+/** Returns whether parameter is in Request object
+ * @param request Request instance
+ * @param param parameter name
+ * @param has whether parameter exists in request
+ * @return int Error code
+ */
+int metkit_marsrequest_has_param(const metkit_marsrequest_t* request, const char* param, bool* has);
+
+/** Returns ParamIterator of parameters in request. Resulting ParamIterator
+ * must be deallocated with metkit_free_paramiterator
+ * @param request Request instance
+ * @param params Allocates ParamIterator object for parameter names in request
+ * @return int Error code
+ */
+int metkit_marsrequest_params(const metkit_marsrequest_t* request, metkit_paramiterator_t** params);
+
+/** Returns number of values for specific parameter in Request object
+ * @param request Request instance
+ * @param param parameter name in request
+ * @param count number of values for param in request
+ * @return int Error code
+ */
+int metkit_marsrequest_count_values(const metkit_marsrequest_t* request, const char* param, size_t* count);
+
+/** Returns value for specific parameter and index in Request object
+ * @param request Request instance
+ * @param param parameter name in request
+ * @param index index of value to retrieve for param in request
+ * @param value retrieved value
+ * @return int Error code
+ */
+int metkit_marsrequest_value(const metkit_marsrequest_t* request, const char* param, int index, const char** value);
+
+/** Returns values for specific parameter Request object
+ * @param request Request instance
+ * @param param parameter name in request
+ * @param values array of values for param in request
+ * @param numValues number of values for param in request
+ * @return int Error code
+ */
+int metkit_marsrequest_values(const metkit_marsrequest_t* request, const char* param, const char** values[], size_t* numValues);
```

Comment:
Does this allocate an array of values in C and then return it? Who is responsible for cleaning up that array of values? Why is this done in a different way to accessing the parameters?

Reply:
Freeing 2d allocations in C is a bit verbose. I suggest either create an opaque type and implement the required query functions or use a public structure that can be returned as handle and has a corresponding free function.

---
Opaque type
```
struct marsrequest_values_t;
typedef marsrequest_values_t* marsrequest_values_handle;
size_t marsrequest_values_size(marsrequest_values_handle h);
const char* marsrequest_values_size(size_t idx);
void marsrequest_values_free(marsrequest_values_handle h);
```

---
Public type
```
typedef struct {
    size_t size;
    const char** values
} marsrequest_values;

void marsrequest_values_free(marsrequest_values* values);
```
---
I think an opaque type would be nice as this could simply be a `std::vector<string>` on the C++ side.